server/supernova/audio_backend/portaudio_backend.hpp

```diff
@@ -121,15 +101,43 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
         }
     }
 
-    bool open_stream(std::string const& input_device, unsigned int inchans, std::string const& output_device,
-                     unsigned int outchans, unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
+    bool open_stream(std::string& input_device, unsigned int inchans, std::string& output_device, unsigned int outchans,
+                     unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
         int input_device_index, output_device_index;
-        if (!match_device(input_device, input_device_index) || !match_device(output_device, output_device_index))
+
+        input_device_index = GetPaDeviceFromName(input_device.c_str(), true);
+        output_device_index = GetPaDeviceFromName(output_device.c_str(), false);
+
+        // check device selection, select defaults if needed
+        fprintf(stdout, "\n");
+        PaError checked =
+            TryGetDefaultPaDevices(&input_device_index, &output_device_index, inchans, outchans, samplerate);
+        fprintf(stdout, "\n");
+
+        // if we got an error from CheckPaDevices, stop here
+        if (checked != paNoError) {
+            report_error(checked);
             return false;
+        }
+
+        // set device names
+        if (inchans)
+            input_device.assign(GetPaDeviceName(input_device_index));
+        if (outchans)
+            output_device.assign(GetPaDeviceName(output_device_index));
```

Comment:
why are these lines conditional on inchans and outchans?

Reply:
setting inchans/outchans to 0 should disable input/output, AFAIU